main/2023/CG-10.md

```diff
@@ -113,4 +113,1850 @@ Use this space to add any agenda items that need to be scheduled, and scheduling
+
+AR: Like a supertype? I don’t have a good use case for that. Maybe that exists in some systems but honestly I don’t know.
+
+CW: That’s all I was hoping to hear
+
+AR: Not sure if that adds any significant complexity. It’s just subtype checking and you can restrict the syntax on one side of that check or not, other than that I don’t think it has one real implication on implementations.
+
+FM: You can’t compile an offset to a structure.. Maybe too specific
+
+AR: What I meant is there is no implication for engines to allow this but disallowing that might be a nice example of excluding some things, knowing you have a common header for structs that define a different module without having to know the extent of all of them. That’s a good example. Might be worth just doing it, but Post-Post-MVP.
+
+<AR Presenting>
+
+TL: There’s one thorny but trivial problem, the type section comes before the import section
+
+AR: Yes, I should have that on the next slide. There is also one minor thing, we want to allow type definitions for that because someone might want to include i31 as their implementation for something.
+
+<AR presenting Type Imports slide>
+
+RT: For now when you cast to an imported type, so you can’t precompile the type during instantiation
+
+AR: Will get back to that problem as well. You are right. We have to think about casts as well in that context.
+
+<AR presenting Type Imports slide>
+
+TL: Of course if you allow interleaving sections, that won't help the case where the import section has to come in the middle of a recursion group.
+
+AR: That won’t work anyway if you have mutual recursion, you can’t import anyway. That’s one thing about isorecursive types, you have to import them together
+
+<AR Presenting Type parameters>
+
+TL: Are these so similar that every function that takes a type import is exactly the same thing? 
+
+AR: that’s the invariant I would want, a little module with an import that can access it’s surroundings so that it has a closure of course. Other than that, that’s the property I want. Efficient way that doesn’t require instantiation for each call.
+
+BT: I would like to see type parameters decoupled from type imports
+
+AR: No, as long as we keep the constraints separate
+
+BT: Seems like more engine work
+
+AR: that’s the point right this is pure validation time. Nothing in the interpreter and I think that’s an important goal to achieve. That’s why I want to keep it separate from value type instantiation.
+
+FM: Before we go into that we should have a proper requirement or use case for it, which group of users are we unblocking? 
+
+AR: I think basically almost every language that has some form of generics except Java generics because they are an afterthought they are compiled away. Almost every other language that has generics has cases where if you don’t jhave that, you have to do this Uppercast. OCaml is one an example, C# is another.
```

Comment:
```suggestion
AR: I think basically almost every language that has some form of generics except Java generics because they are an afterthought they are compiled away. Almost every other language that has generics has cases where if you don’t have that, you have to do this upcast/downcast dance. OCaml is one an example, C# is another.
```